src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -0,0 +1,313 @@
+            var depsFile = new FileInfo(Path.Combine(AppContext.BaseDirectory, depsFileName));
+            if (!depsFile.Exists)
+            {
+                throw new InvalidOperationException(Resources.FormatMissingDepsFile(
+                    depsFile.FullName,
+                    Path.GetFileName(depsFile.FullName)));
+            }
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IWebHostBuilder"/> used to setup <see cref="TestServer"/>.
+        /// <remarks>
+        /// The default implementation of this method looks for a <c>public static IWebHostBuilder CreateDefaultBuilder(string[] args)</c>
+        /// method defined on the entry point of the assembly of <typeparamref name="TEntryPoint" /> and invokes it passing an empty string
+        /// array as arguments. In case this method can't be found,
+        /// </remarks>
+        /// </summary>
+        /// <returns>A <see cref="IWebHostBuilder"/> instance.</returns>
+        protected virtual IWebHostBuilder CreateWebHostBuilder() =>
+            WebHostBuilderFactory.CreateFromTypesAssemblyEntryPoint<TEntryPoint>(Array.Empty<string>()) ??
+            throw new InvalidOperationException($"No method 'public static {nameof(IWebHostBuilder)} CreateWebHostBuilder(string[] args)' " +
+                $"found on '{typeof(TEntryPoint).Assembly.EntryPoint.DeclaringType.FullName}'. Alternatively, " +
+                $"{typeof(WebApplicationFactory<TEntryPoint>).Name} can be extended and 'protected virtual {nameof(IWebHostBuilder)} " +
+                $"{nameof(CreateWebHostBuilder)}()' can be overriden to provide your own {nameof(IWebHostBuilder)} instance.");
+
+        /// <summary>
+        /// Creates the <see cref="TestServer"/> with the bootstrapped application in <paramref name="builder"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> used to
+        /// create the server.</param>
+        /// <returns>The <see cref="TestServer"/> with the bootstrapped application.</returns>
+        protected virtual TestServer CreateServer(IWebHostBuilder builder) => new TestServer(builder);
+
+        /// <summary>
+        /// Gives a fixture an opportunity to configure the application before it gets built.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> for the application.</param>
+        protected virtual void ConfigureWebHost(IWebHostBuilder builder)
+        {
+        }
+
+        private void EnsureBuilder()
+        {
+            if (_builder != null)
+            {
+                return;
+            }
+
+            _builder = CreateWebHostBuilder();
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="HttpClient"/> that automatically follows
+        /// redirects and handles cookies.
+        /// </summary>
+        /// <returns></returns>
+        public HttpClient CreateStandardClient() =>
```

Comment:
How about rename this to `CreateClient` so that people think it's nicer, and rename the empty one to be something like `CreateEmptyClient`?

But, I still have the concern that we discussed last week. The list of handlers we add might grow/change over time, and that would certainly be a breaking change. Can we come up with a way to resolve this?